ChronoKeeper/IngestionQueue.h

```diff
@@ -0,0 +1,142 @@
+#ifndef INGESTION_QUEUE_H
+#define INGESTION_QUEUE_H
+
+
+#include <iostream>
+#include <deque>
+#include <unordered_map>
+#include <mutex>
+
+#include "chronolog_types.h"
+
+//
+// IngestionQueue is a funnel into the MemoryDataStore
+// std::deque guarantees O(1) time for addidng elements and resizing 
+// (vector of vectors implementation)
+
+namespace chronolog
+{
+
+typedef std::deque<LogEvent> EventDeque;
+
+class StoryIngestionHandle
+{
+
+public:
+	StoryIngestionHandle( std::mutex & a_mutex, EventDeque & event_deque)
+       		: handleMutex(a_mutex)
+	  	, activeDeque(&event_deque)
+			{}
+	~StoryIngestionHandle() = default;
+
+	void ingestEvent( LogEvent const& logEvent)
+	{   // assume multiple service threads pushing events on ingestionQueue
+	    std::lock_guard<std::mutex> lock(handleMutex); 
+	    activeDeque->push_back(logEvent); 
+	}
+
+	void swapActiveDeque( EventDeque * empty_deque, EventDeque * full_deque)
+	{
+	    if( activeDeque->empty())
+	    {	    return ;       }
+//INNA: check if atomic compare_and_swap will work here
+
+	    std::lock_guard<std::mutex> lock(handleMutex); 
+	    if(  activeDeque->empty() )
+	    {	    return ;       }
+
+	    full_deque = activeDeque;
+	    activeDeque = empty_deque;
+
+	}
+	     
+private:
+
+	std::mutex & handleMutex;   
+	EventDeque * activeDeque;
+};
+
+
+class IngestionQueue
+{
+public:
+	IngestionQueue()
+	{ }
+	~IngestionQueue()
+	{ shutDown();}
+
+void addStoryIngestionHandle( StoryId const& story_id, StoryIngestionHandle * ingestion_handle)
+{
+	std::lock_guard<std::mutex> lock(ingestionQueueMutex);
+	storyIngestionHandles.emplace(story_id,ingestion_handle);
+}
+
+void removeIngestionHandle(StoryId const & story_id)
+{
+	std::lock_guard<std::mutex> lock(ingestionQueueMutex);
+	storyIngestionHandles.erase(story_id);
+}
+
+void ingestLogEvent(LogEvent const& event)
+{
+	auto ingestionHandle_iter = storyIngestionHandles.find(event.storyId);
+	if( ingestionHandle_iter == storyIngestionHandles.end())
+	{
+		std::cout <<" orphan event"<<std::endl;
```

Comment:
Under what circumstance can an event become orphan? Is it something lighting fast record call from the Client after acquisition before Keeper is ready to that particular Story?